include/Introspection.h

```diff
@@ -0,0 +1,1501 @@
+    /// @pre MemberType needs to be a type that exists on ClassType.
+    template<class ClassType, class MemberType, size_t Num>
+    using MemberArrayPtrType = MemberType(ClassType::*)[Num];
+
+    ///////////////////////////////////////////////////////////////////////////////
+    // Method Types Deduction
+
+    /// @brief A deliberately empty setter deduction struct to catch error cases.
+    /// @tparam T Any type, doesn't matter.  This will catch anything not caught by specializations.
+    template<typename T>
+    struct DeduceSetterTypes
+        {  };
+
+    /// @brief A DeduceSetterTypes specialization for a setter member method.
+    /// @tparam Argument The setter parameter type that will be deduced.
+    /// @tparam Object The object type that will be deduced.
+    template<class Argument, class Object>
+    struct DeduceSetterTypes<void(Object::*)(Argument)>
+    {
+        /// @brief The object type to which the setter method belongs.
+        using ObjectType = Object;
+        /// @brief The argument accepted by the setter method.
+        using ArgumentType = Argument;
+        /// @brief The decayed type of Argument, stripped of reference and const qualifiers.
+        using DecayedType = std::decay_t<Argument>;
+    };//DeduceSetterTypes
+
+    /// @brief A DeduceSetterTypes specialization for a non-const reference setter member method.
+    /// @tparam Argument The setter parameter type that will be deduced.
+    /// @tparam Object The object type that will be deduced.
+    template<class Argument, class Object>
+    struct DeduceSetterTypes<Argument&(Object::*)()>
+    {
+        /// @brief The object type to which the setter method belongs.
+        using ObjectType = Object;
+        /// @brief The argument accepted by the setter method.
+        using ArgumentType = Argument;
+        /// @brief The decayed type of Argument, stripped of reference and const qualifiers.
+        using DecayedType = std::decay_t<Argument>;
+    };//DeduceSetterTypes
+
+    /// @brief A DeduceSetterTypes specialization for a direct member pointer.
+    /// @tparam Member The member pointer that will be deduced.
+    /// @tparam Object The object type that will be deduced.
+    template<class Member, class Object>
+    struct DeduceSetterTypes<Member Object::*>
+    {
+        /// @brief The object to which the member belongs.
+        using ObjectType = Object;
+        /// @brief The member type being pointed to.
+        using ArgumentType = Member;
+        /// @brief The decayed type of Member, stripped of reference and const qualifiers.
+        using DecayedType = std::decay_t<Member>;
+    };//DeduceSetterTypes
+
+    /// @brief A deliberately empty getter deduction struct to catch error cases.
+    /// @tparam T Any type, doesn't matter.  This will catch anything not caught by specializations.
+    template<typename T>
+    struct DeduceGetterTypes
+        {  };
```

Comment:
It would give that error if we did what most people do and not have a struct body.  I've added a body so the class is defined, but the type isn't.  We want it to error, and not finding the type being looked for is perfectly valid in this case.  I don't think static asserts would be all that useful here.

Reply:
Additionally, the checks done with the template specialization here are in tandem with the checks in the MemberAccessor class.  Those add more clarity, and if the static asserts aren't triggered there, then I'm fairly sure there won't be an issue here.  Although I can't say I know that 100%.